src/carp.cpp

```diff
@@ -0,0 +1,163 @@
+  Eigen::SparseLU<Eigen::SparseMatrix<double> > premat_solver;
+  premat_solver.compute(premat);
+
+  // Book-keeping variables
+  // Number of iterations stored, total iteration count, number of fusions
+  Eigen::Index path_iter  = 0;
+  Eigen::Index iter       = 0;
+  Eigen::Index nzeros_old = 0;
+  Eigen::Index nzeros_new = 0;
+
+  while( (iter < max_iter) & (nzeros_new < num_edges) ){
+    // Begin iteration - move updated values to "_old" values
+    //
+    // TODO -- Confirm that these use "move semantics" to avoid full copies
+    u_old        = u_new;
+    v_old        = v_new;
+    z_old        = z_new;
+    vZeroIndsold = vZeroIndsnew;
+    nzeros_old   = nzeros_new;
+
+    // U-update
+    // TODO - Document what is happening here
+    Eigen::VectorXd solver_input = DtMatOpv2(rho * v_old - z_old, n, p, IndMat, EOneIndMat, ETwoIndMat);
+    solver_input += x;
+    solver_input /= rho;
+    u_new = premat_solver.solve(solver_input);
+
+    // V-update
+    // TODO - Document what is happening here
+    Eigen::VectorXd prox_argument = DMatOpv2(u_new, p, IndMat, EOneIndMat, ETwoIndMat) + (1/rho)*z_old;
+
+    if(l1){
+      v_new = ProxL1(prox_argument, p, (1/rho) * gamma, weights);
+    } else {
+      v_new = ProxL2(prox_argument, p, (1/rho) * weights * gamma, IndMat);
+    }
+
+    // Z-update
+    // TODO - Document what is happening here
+    z_new = z_old + rho*(DMatOpv2(u_new, p, IndMat, EOneIndMat, ETwoIndMat) - v_new);
+
+    // TODO- Document this check: what are we trying to do here?
+    for(int l = 0; l < num_edges; l++){
+      Eigen::VectorXi v_index = IndMat.row(l);
+      if(extract(v_new, v_index).sum() == 0){
+        vZeroIndsnew(l) = 1;
+      }
+    }
+    nzeros_new = vZeroIndsnew.sum();
+
+    // If we have seen a fusion or are otherwise interested in keeping this iteration,
+    // add values to our storage buffers
+    if( (nzeros_new != nzeros_old) | (iter % keep == 0) ) {
+
+      // Before we can store values, we need to make sure we have enough buffer space
+      if(path_iter >= buffer_size){
+        buffer_size *= 2; // Double our buffer sizes
+        UPath.conservativeResize(UPath.rows(), buffer_size);
+        VPath.conservativeResize(VPath.rows(), buffer_size);
+        gamma_path.conservativeResize(buffer_size);
+        vZeroInds_Path.conservativeResize(vZeroInds_Path.rows(), buffer_size);
+      }
+
+      // Store values
+      UPath.col(path_iter)          = u_new;
+      VPath.col(path_iter)          = v_new;
+      gamma_path(path_iter)         = gamma;
+      vZeroInds_Path.col(path_iter) = vZeroIndsnew;
+
+      path_iter++;
+    }
+
+    iter++;
+    if(iter >= burn_in){
+      gamma *= t;
+    }
+
+    if((iter % CLUSTRVIZ_CHECK_USER_INTERRUPT_RATE) == 0){
+      Rcpp::checkUserInterrupt();
+    }
+  }
+
+  // Now that we are done, we can "drop" unused buffer space before returning to R
+  UPath.conservativeResize(UPath.rows(), path_iter - 1);
```

Comment:
I think the resize here cuts off the last iteration in the path. I tested locally by changing to:

UPath.conservativeReize(UPath.rows(),path_iter);
VPath.conservativeReize(VPath.rows(),path_iter);
gamma_path.conservativeReize(gamma_Path.rows(),path_iter);
vZerodInds_Path.conservativeReize(vZeroInds_Path.rows(),path_iter);

and the CARP tests pass afterwards.

Reply:
I will try to see if this work for CBASS too.

Reply:
A gentleman and a scholar!

As confirmed by adding the lines

```
if(UPath.col(UPath.cols()) != u_new){
    Rcpp::stop("Oh SNAP!"); 
}
```

reveals, I was indeed dropping the last iteration. 

This is a classic off-by-one error: `path_iter` is the zero-based index of the next column we want to insert to, while `conservativeResize` takes the (one-based) number of columns we want to keep. 

If, for example, our algorithm terminated after 100 save-worthy iterations, we would have `path_iter = 100` (because the possible next iteration would be in column 100, *i.e.*, `UPath.col(101)`), so `path_iter - 1 = 99` and we are loosing a column.

Reply:
Addressed in 116227e